### ISSUE TITLE
Some tests should manually select the local engine variant

### ISSUE THREAD

**Issue by zanderso** [flutter/flutter#72368]
See the discussion here https://www.google.com/url?q=https://github.com/flutter/flutter/pull/71721%23issuecomment-743218994&sa=D&ust=1608060250466000&usg=AOvVaw20Z6aS6edMb3Vf-qBQcHH2

@a-siva @mkustermann

@jonahwilliams who did some plumbing around this recently here https://github.com/flutter/flutter/pull/65802

**Comment by annagrin** [comment]
Note: fixing this will enable us to fix the web HHH bot from failing every time the kernel version changes.

**Comment by annagrin** [comment]
Exsmples: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/command_output_test.dart#L21
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/integration.shard/test_driver.dart#L477

**Comment by godofredoc** [comment]
\cc @jmagman  any pointers that you can give us here?

**Comment by jmagman** [comment]
@yjbanov was investigating something similar in https://github.com/flutter/flutter/issues/78072.

**Comment by jmagman** [comment]
Never mind, this is requesting the possibility to `--local-engine=android_debug` instead of `host_debug`, not that the local engine isn't being used at all in some tests.

**Comment by jmagman** [comment]
The original impetus for this seems to be:
> We are seeing issues on HHH configuration of dart-lang/sdk,
flutter/engine and flutter/flutter where the web tests are using new
frontend_server tooling (which expects current kernel version) but is
given kernel file from an older version.

That may be fixed by https://github.com/flutter/flutter/pull/78496.

However if we still need this, the integration tests could have a `getLocalEngineArguments()` for each desired variant like `getLocalEngineAndroidArguments()` that replaces the injected flag values.  Something like:
```dart
List<String> getLocalEngineAndroidArguments() => 
  getLocalEngineArguments().map((String argument) => argument.replaceAll('host_', 'android_')).toList();
```
but something more robust.

**Comment by annagrin** [comment]
@jmagman I was under impression that we needed to set local engine environment variables to make sure the bots run tests with the local engine, as was attempted in https://github.com/flutter/flutter/pull/71721? I think that PR was reverted. Is it no longer needed?

**Comment by jmagman** [comment]
@annagrin You're right, it's not related to #78496.  `web_tool_tests` is calling `pub` without the local engine flags, not `flutter test`.
https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket.appspot.com/8852389365860202592/+/steps/flutter_test_web_tool_tests/0/stdout

**Comment by annagrin** [comment]
@jmagman @godofredoc I would appreciate if someone could take this over from me, as I am not familiar with the flutter test infrastructure enough to make the right call.

**Comment by dcharkes** [comment]
> Note: fixing this will enable us to fix the web HHH bot from failing every time the kernel version changes.

We had another kernel version change 4 days ago: https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket.appspot.com/8851986851676848432/+/steps/flutter_test_tool_tests/0/stdout

cc @johnniwinther

**Comment by annagrin** [comment]
cc @athomas would it be possible to build other flavors of engine to make sure all tests run correctly when pub runs with local engine flag? See @jmagman investigation above.

**Comment by sigmundch** [comment]
Some observations after looking at this more closely and discussing it with @jonahwilliams:
1. Tests that use `getLocalEngineArguments` are in the `integration.shard`. We'll attempt to set the local-engine environment flags only for the `web.shard` to unblock the 3H-web bot on the Dart side.
2. We may not need a big general change to handle the uses of `getLocalEngineArguments`, but a one by one manual review instead. Currently only 12 files (total of 47 unit tests) use it. Many don't need a local engine, for example, so it may be OK to remove the use of the flag in those cases (see https://github.com/flutter/flutter/pull/83282). Individual tests could be changed to ignore the local engine flag or skipped the test if the specified engine is not compatible with the test.

**Comment by sigmundch** [comment]
It's been a while, so trying to refresh my memory here. Since we landed https://github.com/flutter/flutter/pull/83308, we are no longer affected in the web HHH boots (item 1 in my previous comment), so that part is complete.

I am not aware that any work was done to address item 2. However, I'm not aware of anything being blocked by it either. As such, maybe the remaining work here could be categorized as tech debt/test hygiene and prioritized accordingly?

**Comment by matanlurey** [comment]
Got it, thanks!
